design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,256 @@
+
+## Out of Scope
+
+1. Creating sophisticated implementations of service lifecycle management will not be part of this effort e.g. pooling pre-existing dependency resources to streamline load times
+2. Being exhaustive in testing all the possible versions of services like Conjur. Though this would be thorough this would be time consuming and costly. In the future we might consider having test scenarios that are less costly and less representative of production but close enough to it to be of value. An example is running a branch follower against a stable master.
+3. juxtaposer will eventually need to be automated. This is not addressed as part of this effort
+|
+
+# Experience
+
+## Assumptions
+
+We assume that the experience of running automated tests given configured dependency resources (such as Conjur and target services) is decent enough at this point in time to not warrant consideration in the improvements suggested by this document. This means  a user (a developer or Jenkins) is able to run integration tests in an automated fashion and get meaningful feedback on the test outcome.
+
+Moving the lifecycle management of dependency resources behind an opaque API is a task that can be achieved with minimal effort.
+
+## Overview
+
+There is an idea for a test.
+1. A developer writes the test case in as close to declarative a fashion as possible, leveraging preexisting steps and assertions.
+2. A developer specifies the service dependencies for their test case.
+3. A developer runs a simple command to run the tests, this results in either success or failure, in both cases the outcome is reported and any resources used in the test are cleaned up (unless requested to stay). In the case of failure breadcrumbs are provided fo where in the test the failure occurred.
+4. A developer adds the test to CI and pushes the branch got Github
+5. A repetition of 3 occurs except this case it in CI
+
+## Walkthroughs
+
+TBD
+
+# Technical Design
+
+## Architecture
+
+[ Service ] is any software that tests rely on e.g. Conjur, Postgres etc.
+
+[ Service Config ] specifies details about a service e.g. for Conjur `CONJUR_ACCOUNT`, `version` etc. Conjur will have different configuration options to Postgres. These options will need to be documented. It's important for users of the Service Engine to know what they're able to configure on a service. Config also specifies the lifecyle method of the interface (start, stop, clean)
+
+[ Test Runner ] is software that provides a conventional way for defining and running test cases. They typically come with capabilities to measure test metrics, test outcomes and code coverage. This will also allow historical tracking. An example of a test-runner is Cucumber. Test cases can be defined in Gherkin
+
+[ Service Engine ] uses [ Service Config ] to create an instance of a [ Service ] 
```

Comment:
are `service engines` something that we're planning to write to parse a service config and run the defined services?